(ready to merge) - Fractions skill score

nikeethr opened the pull request:
See #188 for details

<strike>I've initially committed my experimental code. Will be moving the numba/numpy implementation to the main scores package.
I've included the rust implementation in rust/
Examples in rust/examples (which also has numba & fast_fss for comparison)
Uses `maturin` build system.

This PR is malleable, depending on reviewer comments I expect the project layout to change...</strike>

See #273 for rust/numba implementation

---

**Update 2024/05/07**
- `rust` and `numba` implementation moved to experimental branch #270 for now
- this PR will now only track the `numpy` implementation

TODO (initial work):
- [x] numpy implementation
- [x] general integration test
- [x] tutorial notebook
- [x] fix linting errors
- [x] bonus: aggregation/dask see notes: https://github.com/nci/scores/issues/188#issuecomment-2097490317
- [x] tidy up docs
- [x] fine-grained tests
- [x] zero-padding option
- [x] initial notebook review

Redesign work:
- [x] address initial minor review comments
- [x] move backend `fss` calculations to `scores/fast/fss/...` (remove from scores/probabilisitic)
- [x] add API interface for fss in `spatial` with `numpy` threshold operators
- [x] add API interface for fss in `spatial` with binary input fields (assert `np.dtype` is `bool`)
- [x] add testing for binary fss
- [x] Bump coverage back to 100%

Tutorial notebooks:
- [x] tutorial for using binary inputs directly (potentially pre-canned using `EventOperator`)
- [x] tutorial for zero-padding
- [x] notebook review by subject matter expert
- [x] incorporate notebook from scientific review.

Final review:
- [x] double check urls/docs (e.g. reference `doi` links instead)
- [x] go through updated checklists: https://github.com/nci/scores/blob/develop/.github/pull_request_template.md

---

**Manual Testing**

I've used the code referenced in #188 (https://github.com/nathan-eize/fss/blob/master/fss_core.py ) to compare  - it looks to be the same code as in the reference paper https://www.researchgate.net/publication/269222763_Fast_calculation_of_the_Fractions_Skill_Score,, except it adds 0-padding instead of trimming.

My implementation throws out partial windows (i.e. it should match the implementation in the paper), mainly because the reference repo weights partial windows the same as full ones, which may not be the same as trimming. However, this still should be investigated - because the effective FSS score in the implementation computes the FSS score effectively excluding the window border, whereas the 0-padded score can be thought to be border-less.

Regardless, in the tests I tried, it matches to 5-6 decimal points.
 
Also - tested it on synthesized data, but would be good to test on more application specific datasets as well.

The notebook is added in this pull request (FSS.ipynb) for further review/manual testing

---

**Update 2024/05/09**

Added support for zero-padding, as an optional flag. (Currently defaults to false). This is useful for FSS that need to be computed over the entire spatial range of the dataset *or if the window sizes chosen are very large) and therefore need to be padded with zeros at the boundaries.

The default zero-padding = False uses the `border` of the dataset as the padding. Which may provide better accuracy as it is "padded" with real information. The trade-off that the score is only accumulated over the range `[lat0 + window_height / 2, lat1 - window_height / 2], [lon0 + window_width / 2, lon1 - window_width / 2]`. On the other hand, zero padding covers the full lat/lon range, with the caveat that it may include partial windows close to the boundaries which are aggregated over less data points.

---

**Update 2024/05/10**

Code design update after conversations with @tennlee 
- We should move the `fss` score to `scores/continuous` since it accepts a continuous field.
- Instead of the hard coded `>` threshold, we should allow for a choice of `[numpy.greater_equal, numpy_greater, numpy.less_equal, numpy.less]` (at least for the `numpy` backend.)
- The FSS backend calls themselves will move to `scores/fast/fss/...` I think this makes sense as the current FSS implementation is a "fast" implementation (when compared to FFT/naive sliding windows).
- A lot of users will actually have a pre-prepared binary field ready and may not necessarily need to do the computation. To this end, we should provide a "binary_fss" option in `scores/categorical`.
- We can also add a section in the tutorial notebook to show how to use the `EventOperator` in the contingency impl to generate the binary field before computing the `binary_fss`.

As such I propose the following code structure:
```sh
/scores/spatial/fss_impl.py  # -> calls the safe numpy backend, "compute_method" is hidden to user
                                # -> will contain an "advanced" api with "compute_method" for 
                                #    advanced users
                                # -> will contain binary fss function for cases where user has prepared binary field(s)
/scores/fast/fss/backend.py  # -> abstract API for backend algorithm pipeline
             .../fss_numpy.py
# ---
# modules below for illustration purposes only, these backends will be checked against appropriate
# dependency requirements
# ---
             .../fss_numba.py
             .../fss_numba_parallel.py
             .../fss_rust.py
```

Also updated TODO list above to reflect this.

tennlee commented:
If I understand correctly, the intention would be to (eventually) merge both the numpy and numba implementations, witholding the rust experimental code. If that's right, please create a separate branch for the two parts of the work so that we can clearly separate them. Otherwise it looks fine. I would suggest moving the Rust work out into another branch which is not part of any merge request. If we want the "conversation" features / code review features offered by the merge request process, I can create a long-lived experimental branch, and you can create merge requests from your work into the experimental branch.

nikeethr commented:
> If I understand correctly, the intention would be to (eventually) merge both the numpy and numba implementations, witholding the rust experimental code. If that's right, please create a separate branch for the two parts of the work so that we can clearly separate them. Otherwise it looks fine. I would suggest moving the Rust work out into another branch which is not part of any merge request. If we want the "conversation" features / code review features offered by the merge request process, I can create a long-lived experimental branch, and you can create merge requests from your work into the experimental branch.

Yup that’s the plan. I’ll split it out as I refactor the numpy parts out and create a new one for numba.

Re: rust. 
It might be good to have a long lived experimental branch similar to develop, that we can merge experimental features against… like you suggest.

tennlee commented:
I have created a branch 268-long-lived-experimental, which can be used for this purpose. Please adjust this MR to track long-lived-experimental, so it's clear that it should be merged to develop. When you are ready, please create new merge requests for the intended inclusions to the develop branch.

If you think code should be merged to develop but not included in the main namespaces, we will also happily create an 'experimental' namespace. This would allow new metrics or approaches to be included in the PyPI package and main documentation, while allowing change and flagging its experimental nature. However, I think in this case the rust work would not go there.

nikeethr commented:
> I have created a branch 268-long-lived-experimental, which can be used for this purpose. Please adjust this MR to track long-lived-experimental...

Thanks @tennlee. I've actually made two other issues to track `numba` (which will merge to develop) and `rust` (which will merge to the long lived branch). I will rebase this one so that it only contains the `numpy` stuff. So I'll keep this one for the main FSS implementation, if that's okay.

nikeethr commented:
@tennlee ok I've cleaned this up to only contain  `numpy` stubs. I'll implement the numpy version back in here shortly.

I've moved everything else here #273 for now.

nikeethr commented:
@tennlee I've moved this out of draft, so it can start being reviewed. I still need to fix up some docstrings, and linting. 

Other than that, have a look at the tutorial notebook and code and let me know if there's any glaring mistakes.

Meanwhile, I'll catch up on the testing/docs the rest of the week.

nikeethr commented:
@nicholasloveday @tennlee, Added notes re: manual testing in description.

nikeethr commented:
After thinking about it, I might also add in the 0-padded option in the FSS computation, in situations where we can't afford to exclude the window border (e.g. very low res data)

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+# "tuple" of 64 bit floats
+f8x3 = np.dtype("f8, f8, f8")
+# Note: `TypeAlias` on variables for python <=3.9
+if TYPE_CHECKING:  # pragma: no cover
```
Current Python testing is on 3.9 and up, so I think you can drop support for earlier versions.

nikeethr replied:
I meant to say 3.9 and below, not sure why I put <=3.9, <3.10 is what I wanted to say.

nikeethr replied:
As an aside, 3.10 has a lot of improvements in the typing libs, any reason not to bump the support to >=3.10?

tennlee replied:
There was a period where Python going from 3.9 to 3.10 broke a bunch of Linux systems and tools which assumed that version numbers only had one decimal place and lexical ordering. It's an older problem, but not that old. I suspect it's probably okay, but that's why. On the other hand, 3.10 was released in October 2021. I haven't had any complaints about version support to date, so I'm just trying to be conservative. I had some other features which also transition at 3.10 which would be nice to be able to focus on. I will raise the question with the other contributors and see what they thing.

nikeethr replied:
Makes sense. I'll just change the comment to <3.10, to reflect the compatibility reason.

tennlee replied:
I did a bit of extra checking on use cases. Python 3.9 remains a requirement unless we're absolutely unable to maintain compatibility in future.

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+    Choice of compute backend for FSS.
+    - Standard
+        - NUMPY (default)
+    - Optimized
```
If only one thing is available, strip out the other options until such time as we want to include them.

nikeethr replied:
I kept it as a scaffolding in case others want to implement particular methods. But I can remove the unimplemented ones.

nikeethr replied:
Will move all backends to `scores/fast/fss` and preserve the choice there.

The exposed method(s) in `scores/continuous/fss` will only contain the default `numpy` backend for now.

nikeethr replied:
I've removed most of the other methods except numba, just so the example logic on how to check for compatiblity is there in `scores.fast.fss`

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+def fss_2d(
+    fcst: xr.DataArray,
+    obs: xr.DataArray,
+    *,  # Force keywords arguments to be keyword-only
```
The "*" should be placed between the positional arguments and the keyword arguments

nikeethr replied:
I want `threshold` and `window` to be explciity keyword arguments so they are not mixed up, but not have defaults. If you prefer I can set them to `None` and check that they are not `None` down the track - but that gives it a false sense of optionality

tennlee replied:
Ah, that makes sense. No change needed then. Carry on!

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+    fcst: xr.DataArray,
+    obs: xr.DataArray,
+    *,  # Force keywords arguments to be keyword-only
+    threshold: np.float64,
```
What is this threshold. Prefer a slightly more descriptive variable name.

nikeethr replied:
Its to bisect the fields into 0s and 1s. In the reference papers it's used in precipitation for example, so I guess `event_threshold` makes sense? 

If I'm understanding correctly - if we want to compute FSS for rainfall events > X mm, X would be the threshold.

tennlee replied:
Yes, event_threshold sounds good. You could maybe even use the event operator from the binary contingency work to produce the event data. That would be a nice piece of consistency, and also allow the provision of more complex event definitions.

nikeethr replied:
I had a look at that, but it seems to support a lot more scope than a simple operation. Maybe I could break it up into a different issue, since I'll have to look at it more carefully.

tennlee replied:
Well, it seems useful at least to allow simple operations like greater-than-or-equal, greater-than, less-than, less-than-or-equal. These can be accessed from the python 'operator' module. Let's review tomorrow whether to use the full event operator class or just allow simple operator specification.

nikeethr replied:
The other reasoning I think I had is that I didn't want to overload the `ufunc` calls with additional function decoration. I don't think python `inlines`  things very well unless its explicit. `numpy` knows to overload the `>` operator, but I'm not sure about custom operators, at the very least it'll have to unpack the wrapped operator.

nikeethr replied:
As discussed - will add `numpy.greater`, and it's siblings as optional threshold operator choices.

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+    *,  # Force keywords arguments to be keyword-only
+    threshold: np.float64,
+    window: Tuple[int, int],
+    spatial_dims: Tuple[str, str],
```
Maybe just an iterable or collection?

tennlee replied:
This is fine, and it's up to you. But if you want to support lists and sets as well as tuples, consider it. In fact, maybe you want a set because a tuple could have the same thing twice.

nikeethr replied:
Hmm, I chose tuple because, in this context it's:
- explicitly 2 fields
- immutable

Set does make sense for the uniqueness property, but I generally would use sets as read-only inputs if they are frozen or wrapped within an instantiated data type.

There seems to be a `frozenset` type I've not tried before. Maybe I could switch to that.

tennlee replied:
I'm happy with it as is, if it's a deliberate choice for a reason.

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+    )
+
+    # gather additional dimensions to aggregate over.
+    dims = utils.gather_dimensions2(
```
This should be set to just gather_dimensions, there is an imminent MR to unify them

nikeethr replied:
cool will do

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+    dask: str = "forbidden",  # see: `xarray.apply_ufunc` for options
+) -> XarrayLike:  # pylint: disable=too-many-locals
+    """
+    Aggregates the Fractions Skill Score (FSS) over 2-D spatial coordinates specified by
```
Should this be "calculates"? Aggregates makes it sound like it's not actually calculating it and I end up a bit confused about whether this is the thing I want to call to get the fractions skill score.

nikeethr replied:
I guess this is what I'm intending to say: "Uses `fss_2d_single_field` to compute the fraction skills score for each 2D spatial field in the DataArray and then aggregates them over the output of `gather_dimensions`"

tennlee replied:
Yes, that's good - better than my suggestion and better than what you had :)

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+    zero_padding: bool = False,
+    reduce_dims: Optional[Iterable[str]] = None,
+    preserve_dims: Optional[Iterable[str]] = None,
+    compute_method: FssComputeMethod = FssComputeMethod.NUMPY,
```
I think the public API shouldn't include the compute method. In my mind they would be separate fuctions instead in different namespaces. e.g. scores.categorical.fss --> numpy version, scores.fast.fss --> numba version. Also, this is sitting in probability right now, but it should probably be in categorical since it operates on events not on continuous values or probability values.

nikeethr replied:
Right, I misunderstood the modules.

Basically this takes a continuous input, applies a threshold to transform it to categorical and then associates a probabilistic score based on accumulated window population populations essentially like the final score is like a Brier score, Happy to take your guidance on where it sits though.

tennlee replied:
The type hinting for the dims should possibly be Optional[FlexibleDimensionTypes]

tennlee replied:
The compute method is probably fine to keep around if it's going to make the implementation much easier, I'm just thinking as I read.

nikeethr replied:
Also as a follow up re: compute method, if we have scores.fast.fss wouldn’t the categorical/probability etc. be obfuscated by having to call it from a different module tree? I would also imagine the “fast” portion may not have to reimplement the whole algorithm - just the parts that may need to go fast (depending on the backend) - e.g. the aggregation logic is trivially parallel with ufunc and might not need numba, numba only needs to care about the FSS computation for a single field (just as an example). I feel like it’s intuitive to have a central point to choose the backend compute method…

tennlee replied:
Let's have a chat tomorrow as well about some of the options here. If the end result is leaving this as-is, that's probably okay, but I think there might be some more options to discuss first.

nikeethr replied:
Similar resolution proposed as: https://github.com/nci/scores/pull/266#discussion_r1596174639

Re: place for the actual score. The backend will sit in `scores/fast/fss/...`. The implementation for contiunous inputs (where the user provides the threshold operator) sits in `scores/continous/fss_impl.py`. The categorical implementation where the user pre-computes the binary field, will sit in `scores/categorical/binary_fss_impl.py`.

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+            - zero_padding = False => inner window
+            - zero_padding = True => outer window with 0 values
+        reduce_dims: Optional set of additional dimensions to aggregate over
+            (mutually exclusive to `reduce_dims`).
```
...mutually exclusive to `preserve_dims`...

nikeethr replied:
😆 good catch

nikeethr commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+        obs: An array of observations (same spatial shape as `fcst`)
+        threshold: A scalar to compare `fcst` and `obs` fields to generate a
+            binary "event" field.
+        window: A pair of positive integers `(height, width)` of the sliding
```
yup, would `window_dims` work?

nikeethr replied:
I did window_size, since I seemed to use that lingo in the description.

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+        raise DimensionError(f"missing spatial dims {spatial_dims} in obs")
+    for s_dim in spatial_dims:
+        if obs[s_dim].shape != fcst[s_dim].shape:
+            raise DimensionError(f"spatial dims {spatial_dims} shapes do not match for fcst and obs")
```
A slightly longer error message might be helpful ... I think maybe "The spatial extent of fcst and obs do not match for the following spatial dimensions: {spatial_dims}

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+
+    # wrapper defined for convenience, since it's too long for a lambda.
+    def fss_wrapper(da_fcst: xr.DataArray, da_obs: xr.DataArray) -> FssDecomposed:
+        fss_backend = FssBackend.get_compute_backend(compute_method)
```
Don't need this if pulling out the compute option

nikeethr replied:
Resolving since I think keeping the backend framework gives it the option of calling things centrally and point to e.g. `scores.fast.fss`  or `scores.experimental.fss` for "advanced" usage.

tennlee commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+        (fcst_sum, obs_sum, diff_sum) = fss_d
+
+    fss = 0.0
+    denom = obs_sum + fcst_sum
```
What happens if you supply two empty grids and this sum is zero?

tennlee replied:
Just so long as it doesn't div by zero and throw an error. Good to know it's in the tests. We should probably do a walkthrough of this implementation at some point, so make some notes somewhere about these choices. Perhaps also put such notes into the module documentation so that they are available to future readers.

nikeethr commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+        reduce_dims=reduce_dims,
+        preserve_dims=preserve_dims,
+    )
+    dims_reduce = set(dims) - set(spatial_dims)
```
“- set(spatial_dims)” handles that

nikeethr replied:
As discussed will add a `warning` here. Because user may be intending to (or accidentally) preserve `spatial_dims` (or subset of) when that's not actually possible for this particular score.

tennlee commented:
Comments on the notebook follow...

- Consider writing mathjax for the formula
- The section "summarise to sanity check..." needs a bit more detail, to explain what the check is for, why it's being done etc. I would also suggest the term "validity check"
- In section 3 ... suggest the words "this is because FSS is invariant to the exact location within the sliding window where the binary fields match, and so only the total count is used"

I will leave further comments later when I can continue the review

nikeethr commented on src/scores/probability/fss_impl.py:
```diff
@@ -0,0 +1,528 @@
+def fss_2d_single_field(
+    fcst: npt.NDArray[np.float64],
+    obs: npt.NDArray[np.float64],
+    *,
```
See: https://github.com/nci/scores/pull/266#discussion_r1595117292 will resolve it similarly depending on how the discussion goes.

tennlee replied:
Thanks. Per previous, carry on :) ...

nicholasloveday commented:
@nikeethr I just got a notification for a recent commit where you moved this to `scores.continuous`. I think that we should create a `scores.spatial` for FSS to go in. Usually FSS fits in the spatial group of methods in the textbooks (and on https://www.cawcr.gov.au/projects/verification/#Methods_for_spatial_forecasts)

nikeethr commented:
> @nikeethr I just got a notification for a recent commit where you moved this to `scores.continuous`. I think that we should create a `scores.spatial` for FSS to go in. Usually FSS fits in the spatial group of methods in the textbooks (and on https://www.cawcr.gov.au/projects/verification/#Methods_for_spatial_forecasts)

Makes sense to me. @tennlee and I were thinking of having one api in categorical which takes in binary input fields, i.e. the user has performed the thresholding already. Do you think we should just put everything in spatial?

Might be worth us discussing it when everyone is available.

nicholasloveday commented:
> Makes sense to me. @tennlee and I were thinking of having one api in categorical which takes in binary input fields, i.e. the user has performed the thresholding already. Do you think we should just put everything in spatial?

I'd be comfortable with it just in spatial. The metrics in categorical don't assume that the data is binary to begin with anyway.

nikeethr commented:
> I'd be comfortable with it just in spatial. The metrics in categorical don't assume that the data is binary to begin with anyway.

Ok done

nicholasloveday commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+    "\n",
+    "By default `fss_2d` does a sliding window on \"full\" windows (i.e. all grid points in the window are interior to the input field). This is different to a convolutional kernel which slides the window from outside the domain, and hence would require zero-padding to avoid boundary conditions. This would result in \"partial\" windows. \n",
+    "\n",
+    "Most Fourier based sliding window algorithms (e.g. Fast Fourier Transform) provide some sort of option for padding. Without going into too much detail, the algorithm used here does not necessitate this as it doesn't deal with the frequency domain. Regardless there are various reasons to include zero-padding in the computation. The option is hence given with the optional `zero_padding` argume ...
```
```suggestion
    "Most Fourier based sliding window algorithms (e.g. Fast Fourier Transform) provide some sort of option for padding. Without going into too much detail, the algorithm used here does not necessitate this as it doesn't deal with the frequency domain. Regardless, there are various reasons to include zero-padding in the computation. The option is hence given with the optional `zero_padding` argument set to `True` as shown below: "
```

nicholasloveday reviewed (commented):
I just had a quick read through of the tutorial - it looks great! I just had a few small suggestions based on things that jumped out at me.

nicholasloveday commented on tests/spatial/test_fss.py:
```diff
@@ -0,0 +1,263 @@
+"""
```
Can you add a test where `fcst` and `obs` have different dims (e.g., fcst has a lead day dim) to ensure that any broadcasting is done correctly?

nikeethr replied:
- I'll add a test for `NaN`s

- `dask` is actually out of scope for this PR. I've put warnings in the docstring, the option is there in the API but mainly "experimental/use at your own risk" for those who want a way to try it. Testing it at a unit test level only makes sense to me if there's an overarching integration test that already tests it's feasibility. I'd prefer to take out the `dask` option rather than testing it at this stage, if you're not keen on just having the warning.

- tests where `fcst` and `obs` have different dimensions are already included implicitly on most tests (via parametrization) that perform aggregation in which case fcst has `lead_time`. But I can make an explicit test if you prefer.

nicholasloveday replied:
Okay that all sounds good. No need for an explicit test for the last point.

nicholasloveday commented on src/scores/spatial/fss_impl.py:
```diff
@@ -0,0 +1,369 @@
+    threshold_operator: Callable = np.greater,
+    compute_method: FssComputeMethod = FssComputeMethod.NUMPY,
+    dask: str = "forbidden",  # see: `xarray.apply_ufunc` for options
+) -> XarrayLike:
```
```suggestion
) -> xr.DataArray:
```

nicholasloveday commented on src/scores/spatial/fss_impl.py:
```diff
@@ -0,0 +1,369 @@
+    compute_method: FssComputeMethod = FssComputeMethod.NUMPY,
+    check_boolean: bool = True,
+    dask: str = "forbidden",  # see: `xarray.apply_ufunc` for options
+) -> XarrayLike:
```
```suggestion
) -> xr.DataArray:
```

nicholasloveday commented on src/scores/spatial/fss_impl.py:
```diff
@@ -0,0 +1,369 @@
+    """
+
+    if check_boolean:
+        if not (fcst.dtype == np.bool_ and obs.dtype == np.bool_):
```
What if there is a NaN in there?

nikeethr replied:
Then it shouldn't be valid, it's underlying type would probably be float. `check_boolean` is supposed to be a rigid condition. Handling "empty" data is actually not trivial, and is beyond the scope of this function.

nicholasloveday commented on src/scores/spatial/fss_impl.py:
```diff
@@ -0,0 +1,369 @@
+            the input arrays.
+
+    References:
+        1. https://journals.ametsoc.org/view/journals/mwre/149/10/MWR-D-18-0106.1.xml#e3
```
Probably should add the original citation throughout https://journals.ametsoc.org/view/journals/mwre/136/1/2007mwr2123.1.xml

nikeethr replied:
Will do

nicholasloveday reviewed (commented):
Great work @nikeethr!

A few initial comments. I haven't looked at everything in detail

nikeethr commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+{
```
Yup, will do.

nikeethr commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+    "\n",
+    "The resulting binary field is summed up to represent the populace (number of ones/\"true\" values in the window).\n",
+    "\n",
+    "The resulting 2-D field of rolling sums represents \"_Integral Image_\" of the respective forecast and obeservation fields, which is then aggregated over all the sliding windows to compute the fractions skill score.\n",
```
Will do

nikeethr commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+   "source": [
+    "**1. Setup** \n",
+    "\n",
+    "First let's create some random data for our forecast and observation fields. Let's also try out a few scenarios:\n",
```
Is `FSS` exclusively for rainfall? If not, then I'm gearing toward a).

nikeethr replied:
Gonna choose a) for now and emphasize that it's synthetic. Technically even rainfall doesn't quite fit a nicely clipped normal curve, so it'll be somewhat synthetic anyway.

nicholasloveday replied:
Okay, I think a) is fine. It's generally used for rainfall, but there's no reason you can't expand it to work with other parameters.

nikeethr commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+   "source": [
+    "**3 Run FSS**\n",
+    "\n",
+    "Since we only have spatial dims we'll be using `scores.spatial.fss_2d_single_field` for this purpose."
```
Will do.

Just writing the potential reason(s) here for reference:
- User can choose their own aggregation method
- User wants more flexibility on how the computations are parallelized.

nikeethr commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+   ],
+   "source": [
+    "import xarray as xr\n",
+    "import itertools\n",
```
Will remove.

nikeethr commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+    "fcst = np.stack(fcst)    \n",
+    "\n",
+    "# we now have the arrays in raw numpy format, we'll need some annotations\n",
+    "# lets convert it to xarray and check that our final dimensions match\n",
```
The functions that take a `xarray.DataArray` input should be `xarray`, the single field option doesn't take an `xarray` input because it doesn't (shouldn't) need to have that dependency.

nikeethr commented on src/scores/spatial/fss_impl.py:
```diff
@@ -0,0 +1,369 @@
+
+        For an exact usage please refer to ``FSS.ipynb`` in the tutorials.
+
+    Raises:
```
It's a softer condition, so it'll raise a warning. Will add that it raises a warning to the docstring.

nikeethr commented on tests/spatial/fss_test_data.py:
```diff
@@ -0,0 +1,74 @@
+    np.random.seed(seed)
+    h = 40
+    w = 60
+    obs = np.random.normal(loc=obs_pdf[0], scale=obs_pdf[1], size=(h, w))
```
Will do

nikeethr replied:
@nicholasloveday thanks for prompting this test.

With `np.mesh` due to indexing reasons, the implementation will end up using a off-by-one window size in the first row/column. This may have also been an issue in the reference implementation, but the zero-padding would have "softened" its impact.

I didn't catch this in the `numba`/`rust` implementations because I used for loops and it was fine.

Surprisingly, this wouldn't have been caught in about 50% of the random sampling tests even though they were using relatively small `60x40 dataset with 10x10 windows`. Regardless, this shows the importance of curated tests in these scores which I chose to be `3x3` windows on a `5x4` matrix, that I could verify with pen and paper.

The fix is relatively simple - when using meshgrid for indexing, we simply insert an extra row/column of zeros before the first row/column - after which we can perform the integral image calculations as per usual. Note: this is different to the "zero-padding" option. Further, because we're using the clipping method for the zero-padding (as per the reference paper - which is somewhat of a shortcut), this needs to be done for that as well.

nikeethr replied:
Ok I added a suite of manual testing, and it all seems fine now, front insertion of a single row/column of zeros seems to have worked.

I think there may have been a slight inaccuracy in the implementation from the initial reference. Regardless, the existing test results that was more true to the reference implementation didn't change by much after this change: `<0.001`. While, edge cases in small windows are now handled properly.

nikeethr commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+    "        description=\"Fractions skill score\",\n",
+    "    ),\n",
+    ")\n",
+    "da"
```
Makes sense to me, it just starts to slow down the notebook if we increase the window size (since it's also part of the CI). I'll have a think about it but I'm gearing towards adding it in when some computational uplift (numba) gets added in.

nikeethr replied:
Similar to above, will resolve in separate issue when the multi window/threshold implementation is more well thought out.

nikeethr commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2329 @@
+    "fss_out = np.zeros((W, T, S))\n",
+    "\n",
+    "# lets iterate over the indices, setting the `multi_index` and `writeonly` flags\n",
+    "with np.nditer(fss_out, flags=['multi_index'], op_flags=['writeonly']) as it:\n",
```
Yes, but not yet. I think it should be addressed in #353 or #269 . Because:
- I'd like a more efficient version of doing it and that requires some thought.
- This PR already is getting quite heavy.

However, if there's a strong requirement (or is probably the most common use-case) for it to be part of the PR I'll consider it.

nikeethr replied:
I'm gearing towards differing this to a later issue if that's okay...

nikeethr commented:
> Great work @nikeethr!
> 
> A few initial comments. I haven't looked at everything in detail

Thanks for your review so far, good attention to detail already!

nikeethr commented:
Peter Steinle has performed the scientific review. Comments:
> Plonked another test (used uniform distributions, and then had 3rd scenario = 1-obs). Has the correct results at window of size 1, i.e. 0 for thresholds >= 0.5. Values for thresholds of 0.1 and 0.3 are about right - should be close to 0.81 and 0.49 respectively. Larger limits also are ok. So think it exhibits all the right behaviour

> 
> An extra comment:
> It is common to use FSS on percentages as well as/instead of thresholds. Is it too late to ask for this to be added (or should it be a new request?)
>  
> Reason I ask is that FSS with percentages can better reflect user interpretations  - especially where forecasters really look at patterns and generally do a bit of bias correction for peak amounts. This is especially true for rainfall, which is the main use of FSS.
>  
> Another common part of FSS is to impose a lower threshold on frequency - i.e. to get a result both obs & model domain-wide frequencies need to have greater than a certain percentage (0.5%, or 1% or some user specified value). When comparing occurrences at just a few points across a domain the FSS doesn't really make much sense.
> 

There are also some comments over a call, that I'll summarize:
- Tested on real data that Pete had prepared and seemed to run fine
- Some ambiguity re: using the single field version (`numpy`), though clarified that `fss_2d` (`xarray`) is the main API call, and the single field version is for more fine-grained control.
- Suggested some improvements to the notebook in showing theoretical score convergence, for a given distribution.
- Happy with the speed of the computation.

Steph-Chong commented on src/scores/spatial/fss_impl.py:
```diff
@@ -0,0 +1,371 @@
+    dask: str = "forbidden",  # see: `xarray.apply_ufunc` for options
+) -> xr.DataArray:
+    """
+    Uses :py:func:`fss_2d_single_field` to compute the fraction skills score for each 2D spatial
```
Is this meant to be either "Fractions Skill Score" or "fractions skill score"? If this instance is meant to be "fraction skills score", then please ignore this feedback.

nikeethr replied:
Thanks, it should be Fraction**s** Skill Score

Steph-Chong commented on src/scores/spatial/fss_impl.py:
```diff
@@ -0,0 +1,371 @@
+            the input arrays.
+
+        DimensionWarning: If ``spatial_dims`` are attempting to be preserved e.g. in ``preserve_dims``
+    References:
```
```suggestion

    References:
```

Is it worth adding an empty line before "References" (might it make it a bit easier to read)?

nikeethr replied:
Yes.

Steph-Chong commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2354 @@
+    "Since we only have spatial dims it is sufficient to use  `scores.spatial.fss_2d_single_field` for this purpose.\n",
+    "\n",
+    "There are reasons why a user may prefer to use this API:\n",
+    "- The user can choose their own aggregation method\n",
```
Consider adding an empty line before the dot points, so the dot points render properly in readthedocs

Steph-Chong commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2354 @@
+    "The implementation details are beyond the scope of this tutorial please refer to, [Fast calculation of the Fractions Skill Score][fss_ref] for more info\n",
+    "\n",
+    "In summary, computation of a single field requires the following parameters:\n",
+    "- forecast 2-D field (in spatial domain)\n",
```
Consider adding an empty line before the dot points, so the dot points render properly in readthedocs.

Steph-Chong commented on tutorials/FSS.ipynb:
```diff
@@ -0,0 +1,2354 @@
+    "$$\n",
+    "\n",
+    "where,\n",
+    "- $p_o$ is the observation populace > threshold, in one window\n",
```
Consider adding an empty line before the dot points, so the dot points renders properly in readthedocs.

Steph-Chong reviewed (commented):
A few minor things I noticed.

I also noticed a few files where "FSS" or "fss" isn't defined when first used (e.g. "Fractions Skill Score (FSS)"). I don't have the domain expertise to know when it is and isn't required to define the acronym, so wasn't sure whether to make a comment in each instance.

nikeethr commented:
> I also noticed a few files where "FSS" or "fss" isn't defined when first used (e.g. "Fractions Skill Score (FSS)").

It should probably be defined when first used, wherever possible. I'll fix that up.

nikeethr commented:
@tennlee I think I've done most of what's needed for the feature. Could you have a look and see if its good to merge to develop?

I've added Pete's suggestions in the "Things to try next" section, to avoid notebook bloat. They are also applicative in nature and would make more sense to have in the notebook, after we address #353 in a future PR.

nicholasloveday commented:
@nikeethr, can you add it to `docs/included.md` and the table in the README?  I think you can create a new category for spatial methods in the table

nikeethr commented:
> @nikeethr, can you add it to `docs/included.md` and the table in the README? I think you can create a new category for spatial methods in the table

@tennlee mentioned he will add it as part of a different update:

> ... However, just skip adding it to included.md in the docs, because I'll add something new for the score.

tennlee commented on src/scores/utils.py:
```diff
@@ -39,13 +40,139 @@
 
 
-class DimensionError(Exception):
+class DimensionError(ValueError):
```
Why should this inherit from ValueError? I remember trying to think of the right thing to do about this class, I'd be interested in your thinking.

nikeethr replied:
It just adds an intermediate layer. E.g. if you wanted to catch all `ValueError`s but not all `Exception`s.

tennlee commented on src/scores/utils.py:
```diff
@@ -39,13 +40,139 @@
+
+class DimensionWarning(UserWarning):
+    """
+    Custom warning against invalid dimension inputs that, while do not directly
```
The grammar could be clearer here, and perhaps add an example?

Steph-Chong commented on src/scores/spatial/fss_impl.py:
```diff
@@ -0,0 +1,374 @@
+        DimensionWarning: If ``spatial_dims`` are attempting to be preserved e.g. in ``preserve_dims``
+
+    References:
+        1. https://doi.org/10.1175/2007MWR2123.1
```
I am a big fan of DOIs, as they are more stable and often resolve better than other links!

I would suggest, in addition to including DOIs, to cite the full reference. For instance:

Roberts, N. M., & Lean, H. W. (2008). Scale-Selective Verification of Rainfall Accumulations from High-Resolution Forecasts of Convective Events. Monthly Weather Review, 136(1), 78-97. https://doi.org/10.1175/2007MWR2123.1

I like having the full reference, as it allows readers to see useful information at a glance (eg. the authors, or the year or publication etc.). Plus, if the DOI is having a bad day and is choosing not to resolve on any given day, if you have the full reference you can then search for it and find it via an alternative link. 

My personal preference is also to include the full journal name (and not just journal title abbreviations) - but I don't know that that bothers everyone else :-).

tennlee commented:
I have pushed a few tweaks into this PR to improve the documentation rendering. Thanks for a huge amount of work in preparing this code. There are a few more 'finishing tasks' which will still need to be done, but this is a practical integration point and then we can continue to work on the remaining items without having to keep such a large PR open and managing the risk of the branches diverging over time.
